### PR TITLE
feat: add the ability to specify a custom decode function

### LINKED ISSUE #717: Support custom text decoding function
I've been working on [a text editor](https://github.com/lukedan/codepad) that handles raw binary data of arbitrary character encoding, and I'm trying to integrate tree-sitter into it for syntax highlighting. However, it seems that tree-sitter only supports UTF-8 and UTF-16. I did a bit of digging and discovered that, in `ts_lexer__get_lookahead` in [lexer.c](https://github.com/tree-sitter/tree-sitter/tree/master/lib/src/lexer.c), a `UnicodeDecoderFunction` is used to decode text which is actually general enough to be extended to a lot of other encodings. It would be great (and probably fairly simple) to expose this as part of the interface instead, so that users can use whichever encoding they want by providing custom decode functions.

### PR DESCRIPTION
**Pull request by amaanq**
Closes #717

### CODE REVIEW

**Comment by amaanq** [comment]
Some tests could be useful, but I wonder what encodings we could/should test on

**Comment by ahlinc** [comment]
An approach with a global pointer to a decoding function doesn't allow to work with different parser instances and files in various encodings in a multi threaded fashion.

**Comment by amaanq** [comment]
> An approach with a global pointer to a decoding function doesn't allow to work with different parser instances and ~files~ input streams with various custom encodings in a multi threaded fashion.

Yep, I know. I wasn't sure if it would be better to have it be global or not (I more or less followed the approach of the tree-sitter allocator)

**Comment by ahlinc** [comment]
> > An approach with a global pointer to a decoding function doesn't allow to work with different parser instances and ~files~ input streams with various custom encodings in a multi threaded fashion.
> 
> Yep, I know. I wasn't sure if it would be better to have it be global or not (I more or less followed the approach of the tree-sitter allocator)

Allocator is a different story

**Comment by amaanq** [comment]
reworked it to be local to the parse instance - but the Rust side seems a bit rough to implement :sweat_smile:

**Review by soz-ai — COMMENTED** [review]
[]()

**Comment by amaanq** [comment]
implemented the rust side and tested it on some utf-32 and windows cp1252 😁

File: lib/src/parser.c [hunk 1]
```diff
@@ -1990,21 +1991,23 @@ TSTree *ts_parser_parse_string(
...
 
 TSTree *ts_parser_parse_string_encoding(
   TSParser *self,
   const TSTree *old_tree,
   const char *string,
   uint32_t length,
-  TSInputEncoding encoding
+  TSInputEncoding encoding,
+  UnicodeDecodeFunction decode_function
 ) {
```
Comment (theHamsta) [inline lib/src/parser.c:2004]: Isn't this a breaking change that requires increasing the minor version?

**Comment by ahlinc** [comment]
After reviewing this again I think this may be implemented in a less intrusive way:
- First of all keep all existing C APIs untouched.
- An addition of `TSInputEncoding_TSInputEncodingCustom` doesn't needed at all or it may be made hidden.
- Introduce a one `ts_parser_parse_decode` C API function that would receive a pointer to a decoding fn and store it in the `payload` field of `TSInput` struct so the last may be kept unchanged. And set the encoding field to a hidden sentinel.
- The decoding fn may actually return not a buffer of bytes but a buffer of u32 decoded code points and an internal `decode` function would be noop in this case.

**Comment by amaanq** [comment]
Supersedes by #3833

**Pull request closed without merging**